src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java

```diff
@@ -1752,6 +1753,53 @@ protected Boolean defaultAction(DocTree node, Content content) {
         return result;
     }
 
+    private boolean equalsIgnoreCase(Name name, String s) {
+        return name != null && name.toString().equalsIgnoreCase(s);
+    }
+
+    private boolean hasIdAttribute(StartElementTree node) {
+        return node.getAttributes().stream().anyMatch(
+                dt -> equalsIgnoreCase(((AttributeTree)dt).getName(), "id"));
+    }
+
+    private void generateHeaderId(StartElementTree node, List<? extends DocTree> trees, Content content) {
+        StringBuilder sb = new StringBuilder();
+        String tagName = node.getName().toString().toLowerCase(Locale.ROOT);
+        for (DocTree docTree : trees.subList(trees.indexOf(node) + 1, trees.size())) {
+            if (docTree instanceof TextTree text) {
+                sb.append(text.getBody());
+            } else if (docTree instanceof LiteralTree literal) {
+                sb.append(literal.getBody().getBody());
+            } else if (docTree instanceof LinkTree link) {
+                var label = link.getLabel();
+                sb.append(label.isEmpty() ? link.getReference().getSignature() : label.toString());
+            } else if (docTree instanceof EndElementTree endElement
+                    && equalsIgnoreCase(endElement.getName(), tagName)) {
+                break;
+            } else if (docTree instanceof StartElementTree nested
+                    && equalsIgnoreCase(nested.getName(), "a")
+                    && hasIdAttribute(nested)) {
+                return; // Avoid generating id if embedded <a id=...> is present
+            }
+        }
+        String idValue = sb.toString()
+                .toLowerCase(Locale.ROOT)
+                .trim()
+                .replaceAll("[^\\w_-]+", "-");
+        if (!idValue.isEmpty()) {
+            // Make id unique
+            idValue = idValue + "-hdr";
+            if (!headerIds.add(idValue)) {
+                int counter = 1;
+                while (!headerIds.add(idValue + counter)) {
+                    counter++;
+                }
+                idValue = idValue + counter;
+            }
+            content.add("id=\"").add(idValue).add("\"");
+        }
```

Comment:
This part of the algorithm, to create an `id` from the text contents of a header, still feels like it belongs in `HtmlIds`.  It feels like we should have 
```
HtmlIds.getIdForHeader(CharSequence headerText)
```
or something like that

Reply:
I moved the code to a new `HtmlIds.forHeading(CharSequence, Set<String>)` method. The `forHeading` follows the naming scheme of `HtmlIds`. The second argument of type `Set<String>` is to make the returned id unique within its page.